src/prng/isaac.rs

```diff
@@ -198,44 +203,50 @@ impl IsaacRng {
 
         self.a = a;
         self.b = b;
-        self.cnt = RAND_SIZE as u32;
+        self.index = 0;
     }
 }
 
 impl Rng for IsaacRng {
     #[inline]
     fn next_u32(&mut self) -> u32 {
-        if self.cnt == 0 {
-            // make some more numbers
+        // Using a local variable for `index`, and checking the size avoids a
+        // bounds check later on.
+        let mut index = self.index as usize;
+        if index >= RAND_SIZE {
             self.isaac();
+            index = 0;
         }
-        self.cnt -= 1;
-
-        // self.cnt is at most RAND_SIZE, but that is before the
-        // subtraction above. We want to index without bounds
-        // checking, but this could lead to incorrect code if someone
-        // misrefactors, so we check, sometimes.
-        //
-        // (Changes here should be reflected in Isaac64Rng.next_u64.)
-        debug_assert!((self.cnt as usize) < RAND_SIZE);
-
-        // (the % is cheaply telling the optimiser that we're always
-        // in bounds, without unsafe. NB. this is a power of two, so
-        // it optimises to a bitwise mask).
-        self.rsl[self.cnt as usize % RAND_SIZE].0
+
+        let value = self.rsl[index];
```

Comment:
Does the `index % RAND_SIZE` trick used elsewhere improve benchmarks?

It may not since the optimiser may already be able to infer the bounds on `index`.

Reply:
I did a lot of measuring of this code, and also glanced over the assembly.

We already have to check the size of `index` to see whether a new block of random numbers has to be generated. If the comparison uses `>=` instead of `==`, the bounds check later can be optimised out. It has to use a local variable for `index` though, because when `self.isaac*()` is not inlined the optimiser can not see that the index gets reset.

So the bounds check is already optimised out, and the mask makes it slower (a couple of percent if I remember right).